parachain/pallets/control/src/lib.rs

```diff
@@ -166,41 +200,47 @@ pub mod pallet {
 		) -> DispatchResult {
 			ensure_root(origin)?;
 
-			let params_hash = initializer.as_ref().map(|i| T::MessageHasher::hash(i.params.as_ref()));
-
-			let message = Message {
-				origin: T::OwnParaId::get(),
-				command: Command::Upgrade { impl_address, impl_code_hash, initializer },
+			let initializer_params_hash = initializer.as_ref().map(|i| T::MessageHasher::hash(i.params.as_ref()));
+			let command = Command::Upgrade {
+				impl_address,
+				impl_code_hash,
+				initializer
 			};
-			Self::submit_outbound(message)?;
+			Self::send(T::OwnParaId::get(), command)?;
 
-			Self::deposit_event(Event::<T>::Upgrade { impl_address, impl_code_hash, params_hash });
+			Self::deposit_event(Event::<T>::Upgrade { impl_address, impl_code_hash, initializer_params_hash });
 			Ok(())
 		}
 
-		/// Sends a message to the Gateway contract to create a new Agent representing `origin`
+		/// Sends a command to the Gateway contract to instantiate a new agent contract representing `origin`.
 		///
-		/// - `origin`: Must be `MultiLocation`
+		/// There are two modes of operation, depending on the relative ancestry of the origin:
+		///
+		/// If the origin is a sibling parachain, the command will be sent over the BridgeHub's own channel
+		/// to the Gateway. The sibling will also be charged an upfront fee `T::Fee`, which will cover the
+		/// cost of execution on Ethereum.
+		///
+		/// If the origin is a child of a sibling parachain, then the command will be sent over channel of
+		/// the sibling containing the child. The sibling will assume the cost of execution on Ethereum.
+		///
+		/// - `origin`: Must be `MultiLocation` of a sibling parachain or a child of a sibling parachain
 		#[pallet::call_index(1)]
 		#[pallet::weight(T::WeightInfo::create_agent())]
 		pub fn create_agent(origin: OriginFor<T>) -> DispatchResult {
-			let origin_location: MultiLocation = T::AgentOrigin::ensure_origin(origin)?;
-
-			Self::charge_fee(&origin_location)?;
+			let origin_location: MultiLocation = T::AgentOwnerOrigin::ensure_origin(origin)?;
 
-			let OriginInfo { reanchored_location, agent_id, .. } =
-				Self::process_origin_location(&origin_location)?;
+			// Ensure that origin location is some consensus system on a sibling parachain
+			let (para_id, agent_id, ancestry) = ensure_sibling_or_sibling_child::<T>(&origin_location)?;
```

Comment:
Nice to remove the reanchor logic.

Reply:
yeah we don't really need to reanchor anymore. It seems that was only required in order to support local locations within BridgeHub. However we only need to support locations on sibling parachains.